### PR TITLE
Fortran bindings

### PR DESCRIPTION
**Pull request by angainor**
WIP. At this stage the low-level API is exported, and the original transport layer benchmarks are working (UCX)

### CODE REVIEW

File: benchmarks/transport/ghex_p2p_bi_cb_avail_mt.cpp [hunk 1]
```diff
@@ -79,15 +79,7 @@ int main(int argc, char *argv[])
     }
 #endif
 
-#ifdef USE_OPENMP
```
Comment (mbianco) [inline benchmarks/transport/ghex_p2p_bi_cb_avail_mt.cpp]: Why are these removed? Is the benchmark explicitly not using OpenMP? (also for the other similar changes)

File: benchmarks/transport/mpi_p2p_bi_avail_mt.cpp [hunk 2]
```diff
@@ -3,10 +3,38 @@
...
 
-#include <ghex/transport_layer/ucx/threads.hpp>
 #include <ghex/common/timer.hpp>
 #include "utils.hpp"
 
+#ifdef USE_OPENMP
+#include <omp.h>
+#define DO_PRAGMA(x) _Pragma(#x)
+
+#define GET_THREAD_NUM()  omp_get_thread_num()
```
Comment (mbianco) [inline benchmarks/transport/mpi_p2p_bi_avail_mt.cpp]: This seems to be only valid for OpenMP. Is this test only available for this, or should the code use thread-primitives?

Comment (angainor) [reply benchmarks/transport/mpi_p2p_bi_avail_mt.cpp]: this is only for this benchmark, no need to make it general.

File: benchmarks/transport/ghex_p2p_bi_cb_avail_mt.cpp (hunk 1 above)
Comment (angainor) [reply benchmarks/transport/ghex_p2p_bi_cb_avail_mt.cpp]: This has been simply moved and cleaned up. I use the primitives, and the OMP check is moved to the front of the file.

File: benchmarks/transport/fhex_p2p_bi_cb_avail_mt.f90 [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+    type(ghex_message), dimension(:), allocatable, save :: smsgs, rmsgs
+    type(ghex_request), dimension(:), allocatable, save :: sreqs, rreqs
+    type(ghex_future), save :: bsreq, brreq
+    type(ghex_message), save :: bsmsg, brmsg
+    logical :: result = .false.
+    real :: ttic = 0, tic = 0, toc = 0
+  
+    procedure(f_callback), pointer, save :: rcb, scb
+
+    !$omp threadprivate(comm, rank, size, num_threads, peer_rank)
```
[the remark this answers is not in the document]
Comment (angainor) [reply benchmarks/transport/fhex_p2p_bi_cb_avail_mt.f90]: I am not sure, I will need your help with this :) I guess the compute parts could be moved to C/C++, or if the user wants - he could use fortran cuda interface?

File: benchmarks/transport/fhex_p2p_bi_cb_avail_mt.f90 [hunk 4]
```diff
@@ -0,0 +1,340 @@
+PROGRAM fhex_bench
```
[the remark this answers is not in the document]
Comment (angainor) [reply benchmarks/transport/fhex_p2p_bi_cb_avail_mt.f90]: `fhex` is fotran ghex :) that's just a proposal, if you don't like it we can change. The point is a bit that I need an .so lib for the bindings. For now it is called libfhex.

File: benchmarks/transport/CMakeLists.txt [hunk 5]
```diff
@@ -1,11 +1,14 @@
+find_package(OpenMP REQUIRED)
+enable_language(Fortran)
```
[the remark this answers is not in the document]
Comment (angainor) [reply benchmarks/transport/CMakeLists.txt]: true

File: benchmarks/transport/fhex_p2p_bi_cb_avail_mt.f90 (hunk 4 above)
[the remark this answers is not in the document]
Comment (mbianco) [reply benchmarks/transport/fhex_p2p_bi_cb_avail_mt.f90]: It's fine, I was commenting on the tokens `p2p`, `bi`, `cb`, `avail` and `mt`. It's quite a bit to decode there

File: benchmarks/transport/fhex_p2p_bi_cb_avail_mt.f90 (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbianco) [reply benchmarks/transport/fhex_p2p_bi_cb_avail_mt.f90]: We have experience in dealing with part of the computations doing OpenACC and part C++/CUDA. I think something could be possible also with OpenMP for acceletators.

File: benchmarks/transport/CMakeLists.txt [hunk 6]
```diff
@@ -33,3 +33,23 @@ if (GHEX_USE_UCP)
         endforeach()
     endif()
 endif()
+
+if (GHEX_BUILD_FORTRAN)
+    enable_language(Fortran)
```
Comment (boeschf) [inline benchmarks/transport/CMakeLists.txt:38]: is the `enable_language` needed (since it is already done in the main CMakeLists.txt)?

File: bindings/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,18 @@
+enable_language(Fortran)
```
Comment (boeschf) [inline bindings/CMakeLists.txt:1]: `enable_language` again - this should be already enabled, shouldn't it?

File: bindings/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+find_package(MPI REQUIRED)
```
Comment (boeschf) [inline bindings/CMakeLists.txt:2]: MPI should have been found before

File: include/ghex/transport_layer/ucx/endpoint.hpp [hunk 9]
```diff
@@ -64,7 +64,7 @@ namespace gridtools {
                     void destroy()
                     {
                         if (!m_moved){
-                            ucs_status_ptr_t ret = ucp_ep_close_nb(m_ep, UCP_EP_CLOSE_MODE_FLUSH);
+                            ucs_status_ptr_t ret = ucp_ep_close_nb(m_ep, UCP_EP_CLOSE_MODE_FORCE);
```
Comment (boeschf) [inline include/ghex/transport_layer/ucx/endpoint.hpp:67]: I thought I've fixed this? please merge with master

File: tests/CMakeLists.txt [hunk 10]
```diff
@@ -126,7 +126,7 @@ foreach(_var ${_variants})
     endif()
 endforeach(_var)
 
-set(_tests_rma local_rma simple_regular_exchange)
```
Comment (boeschf) [inline tests/CMakeLists.txt:129]: why is this test gone?

Comment (angainor) [reply tests/CMakeLists.txt:129]: https://github.com/GridTools/GHEX/pull/48/commits/5b0986ac7a782f2de423da752c790aa8aa0cf36e

This is due to my incomplete cherry-pick. You have to re-add it after merging with benchmarks.

File: include/ghex/transport_layer/ucx/endpoint.hpp (hunk 9 above)
Comment (angainor) [reply include/ghex/transport_layer/ucx/endpoint.hpp:67]: I am merged with master. No idea why this is here. Is your other changes (the barrier) present?

File: bindings/CMakeLists.txt (hunk 8 above)
Comment (angainor) [reply bindings/CMakeLists.txt:2]: not really - it is found before `enable_language(Fortran)`, so `MPI::MPI_Fortran` is not available. Would have to change main `CMakeLists.txt`. OK?

File: include/ghex/transport_layer/ucx/endpoint.hpp (hunk 9 above)
Comment (boeschf) [reply include/ghex/transport_layer/ucx/endpoint.hpp:67]: the barrier is there, very strange. Could you still change it back to flush?

Comment (angainor) [reply include/ghex/transport_layer/ucx/endpoint.hpp:67]: This is very strange. It is the only line, which did not get merged. All your other changes from f237b72741b20efad84e18720b65c5ea7f53d5e9 are in. Also, blame shows that my version is much older. So I do not understand why it wasn't merged.

File: bindings/CMakeLists.txt (hunk 8 above)
Comment (angainor) [reply bindings/CMakeLists.txt]: I moved `enable_language(Fortran)` to the beginning of main `CMakeLists.txt`, together with CUDA, and before `find_package(MPI)`

File: include/ghex/transport_layer/ucx/endpoint.hpp (hunk 9 above)
Comment (angainor) [reply include/ghex/transport_layer/ucx/endpoint.hpp:67]: This is very scary. Do we have to understand this? What else is missing?

File: bindings/obj_wrapper.hpp [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+                };
+	
+                std::unique_ptr<obj_type_info> m_obj_storage;
+
+                obj_wrapper(obj_wrapper &&) = default;
+
+                template <class Arg, class Decayed = typename std::decay<Arg>::type>
+                obj_wrapper(Arg &&arg) : m_obj_storage(new obj_storage<Decayed>(std::forward<Arg>(arg))) {}
+
+                std::type_info const &type_info() const noexcept { return m_obj_storage->type_info(); }
```
Comment (boeschf) [inline bindings/obj_wrapper.hpp:39]: do we use the `type_info` anywhere? Only `get_object_ptr_unsafe` is implemented below.

File: bindings/context_bind.hpp [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+#include <ghex/transport_layer/ucx/context.hpp>
+using transport    = ghex::tl::ucx_tag;
+#else
+
+/* fallback MPI backend */
+#include <ghex/transport_layer/mpi/context.hpp>
+using transport    = ghex::tl::mpi_tag;
+#endif
+
+using context_type = typename gridtools::ghex::tl::context_factory<transport>::context_type;
```
Comment (boeschf) [inline bindings/context_bind.hpp:22]: is there a reason we don't use a namespace for these types?

File: bindings/context_bind.hpp [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+using context_uptr_type = std::unique_ptr<context_type>;
+using communicator_type = context_type::communicator_type;
+
+extern context_uptr_type context;
+extern int __GHEX_nthreads;
```
Comment (boeschf) [inline bindings/context_bind.hpp:27]: double underscore may be illegal - any reason why we need it?

File: bindings/context_bind.cpp [hunk 14]
```diff
@@ -0,0 +1,51 @@
+#include "context_bind.hpp"
+#include <mpi.h>
+#include <sched.h>
+#include <sys/sysinfo.h>
+
+context_uptr_type context;
+int __GHEX_nthreads = 1;
+
+/* barrier has to be shared between the threads */
+gridtools::ghex::tl::barrier_t *barrier = NULL;
```
Comment (boeschf) [inline bindings/context_bind.cpp:10]:
```suggestion
gridtools::ghex::tl::barrier_t *barrier = nullptr;
```

File: bindings/context_bind.cpp [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    barrier = new gridtools::ghex::tl::barrier_t(nthreads);
+}
+
+extern "C"
+void ghex_finalize()
+{
+    context.reset();
+    delete barrier;
+    barrier = NULL;
+    __GHEX_nthreads = 1;
```
Comment (boeschf) [inline bindings/context_bind.cpp:28]: why needed?
also please use `nullptr` instead of `NULL`

File: bindings/request_bind.hpp [hunk 16]
```diff
@@ -0,0 +1,12 @@
+#ifndef GHEX_FORTRAN_REQUEST_BIND_INCLUDED_HPP
+#define GHEX_FORTRAN_REQUEST_BIND_INCLUDED_HPP
+
+#include <cstdint>
+
+#define GHEX_REQUEST_SIZE 24
+
+struct frequest_type {
+    int8_t data[GHEX_REQUEST_SIZE];
```
[the remark this answers is not in the document]
Comment (angainor) [reply bindings/request_bind.hpp:9]: Not on the Fortran side, and I need to declare the same type there. Hence IMO it's better to have those spelled out explicitly. But I agree, this is a fishy part and if the size doesn't match, we're in trouble. We need at least some CMake tests to verify that.

File: bindings/context_bind.hpp [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+#include <vector>
+#include <ghex/transport_layer/util/barrier.hpp>
+
+namespace ghex = gridtools::ghex;
+
+#ifdef GHEX_USE_UCP
+
+/* UCX backend */
+#include <ghex/transport_layer/ucx/context.hpp>
+using transport    = ghex::tl::ucx_tag;
```
Comment (boeschf) [inline bindings/context_bind.hpp:14]: should we move those types into a namespace as well?

Comment (angainor) [reply bindings/context_bind.hpp:14]: darn! I still left out some! will do

File: bindings/communicator_bind.cpp [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+void ghex_comm_post_send_cb(obj_wrapper *wcomm, ghex::tl::cb::any_message *wmessage, int rank, int tag, f_callback cb, frequest_type *freq, void *user_data)
+{
+    communicator_type *comm = get_object_ptr_unsafe<communicator_type>(wcomm);
+    if(nullptr==wmessage){
+	std::cerr << "ERROR: trying to submit a NULL message in " << __FUNCTION__ << ". Terminating.\n";
+	std::terminate();
+    }
+    auto req = comm->send(*wmessage, rank, tag, callback{cb, user_data});
+    if(!freq) return;
+    new(freq->data) decltype(req)(std::move(req));
```
Comment (boeschf) [inline bindings/communicator_bind.cpp:109]: similar question here: how do we control lifetime of these objects allocated in-place? Since they are type erased, we need type injection to properly destruct them at some point.

File: bindings/communicator_bind.cpp [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+    }
+    auto fut = comm->send(*wmessage, rank, tag);
+    new(ffut->data) decltype(fut)(std::move(fut));
+}
+
+extern "C"
+void ghex_comm_post_send_cb(obj_wrapper *wcomm, ghex::tl::cb::any_message *wmessage, int rank, int tag, f_callback cb, frequest_type *freq, void *user_data)
+{
+    communicator_type *comm = get_object_ptr_unsafe<communicator_type>(wcomm);
+    if(nullptr==wmessage){
```
Comment (boeschf) [inline bindings/communicator_bind.cpp:103]:
```suggestion
    if(!wmessage){
```

File: bindings/communicator_bind.cpp [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+    new(freq->data) decltype(req)(std::move(req));
+}
+
+extern "C"
+void ghex_comm_send_cb(obj_wrapper *wcomm, ghex::tl::cb::any_message **wmessage_ref, int rank, int tag, f_callback cb, frequest_type *freq, void *user_data)
+{
+    ghex::tl::cb::any_message *wmessage;
+    communicator_type *comm = get_object_ptr_unsafe<communicator_type>(wcomm);
+    wmessage = *wmessage_ref;
+    if(nullptr==wmessage){
```
Comment (boeschf) [inline bindings/communicator_bind.cpp:118]:
```suggestion
    if(!wmessage){
```

File: bindings/communicator_bind.cpp [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+
+/*
+   SEND_MULTI requests
+ */
+
+extern "C"
+void ghex_comm_post_send_multi(obj_wrapper *wcomm, ghex::tl::cb::any_message *wmessage, int *ranks, int nranks, int tag, frequest_type *ffut)
+{
+    communicator_type *comm = get_object_ptr_unsafe<communicator_type>(wcomm);
+    if(nullptr==wmessage){
```
Comment (boeschf) [inline bindings/communicator_bind.cpp:137]:
```suggestion
    if(!wmessage){
```

File: bindings/communicator_bind.cpp [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+    ranks_array.assign(ranks, ranks+nranks);
+    auto fut = comm->send_multi(*wmessage, ranks_array, tag);
+    new(ffut->data) decltype(fut)(std::move(fut));
+}
+
+extern "C"
+void ghex_comm_post_send_multi_cb(obj_wrapper *wcomm, ghex::tl::cb::any_message *wmessage, int *ranks, int nranks, int tag, f_callback cb, frequest_type *freq, void *user_data)
+{
+    communicator_type *comm = get_object_ptr_unsafe<communicator_type>(wcomm);
+    if(nullptr==wmessage){
```
Comment (boeschf) [inline bindings/communicator_bind.cpp:151]:
```suggestion
    if(!wmessage){
```

File: bindings/communicator_bind.cpp [hunk 23]
```diff
@@ -0,0 +1,241 @@
...
+
+extern "C"
+void ghex_comm_send_multi_cb(obj_wrapper *wcomm, ghex::tl::cb::any_message **wmessage_ref, int *ranks, int nranks, int tag, f_callback cb, frequest_type *freq, void *user_data)
+{
+    ghex::tl::cb::any_message *wmessage;
+    communicator_type *comm = get_object_ptr_unsafe<communicator_type>(wcomm);
+    std::vector<int> ranks_array(nranks);
+    ranks_array.assign(ranks, ranks+nranks);
+    wmessage = *wmessage_ref;
+    if(nullptr==wmessage){
```
Comment (boeschf) [inline bindings/communicator_bind.cpp:169]:
```suggestion
    if(!wmessage){
```

File: bindings/communicator_bind.cpp [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+
+/*
+   RECV requests
+ */
+
+extern "C"
+void ghex_comm_post_recv(obj_wrapper *wcomm, ghex::tl::cb::any_message *wmessage, int rank, int tag, ffuture_type *ffut)
+{
+    communicator_type *comm = get_object_ptr_unsafe<communicator_type>(wcomm);
+    if(nullptr==wmessage){
```
Comment (boeschf) [inline bindings/communicator_bind.cpp:188]:
```suggestion
    if(!wmessage){
```

File: bindings/communicator_bind.cpp [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+    }
+    auto fut = comm->recv(*wmessage, rank, tag);
+    new(ffut->data) decltype(fut)(std::move(fut));
+}
+
+extern "C"
+void ghex_comm_post_recv_cb(obj_wrapper *wcomm, ghex::tl::cb::any_message *wmessage, int rank, int tag, f_callback cb, frequest_type *freq, void *user_data)
+{
+    communicator_type *comm = get_object_ptr_unsafe<communicator_type>(wcomm);
+    if(nullptr==wmessage){
```
Comment (boeschf) [inline bindings/communicator_bind.cpp:200]:
```suggestion
    if(!wmessage){
```

File: bindings/communicator_bind.cpp [hunk 26]
```diff
@@ -0,0 +1,241 @@
...
+    new(freq->data) decltype(req)(std::move(req));
+}
+
+extern "C"
+void ghex_comm_recv_cb(obj_wrapper *wcomm, ghex::tl::cb::any_message **wmessage_ref, int rank, int tag, f_callback cb, frequest_type *freq, void *user_data)
+{
+    ghex::tl::cb::any_message *wmessage;
+    communicator_type *comm = get_object_ptr_unsafe<communicator_type>(wcomm);
+    wmessage = *wmessage_ref;
+    if(nullptr==wmessage){
```
Comment (boeschf) [inline bindings/communicator_bind.cpp:215]:
```suggestion
    if(!wmessage){
```

File: bindings/communicator_bind.cpp [hunk 27]
```diff
@@ -0,0 +1,241 @@
...
+
+/*
+   resubmission of recv requests from inside callbacks
+ */
+
+extern "C"
+void ghex_comm_resubmit_recv(obj_wrapper *wcomm, ghex::tl::cb::any_message *wmessage, int rank, int tag, f_callback cb, frequest_type *freq, void *user_data)
+{
+    communicator_type *comm = get_object_ptr_unsafe<communicator_type>(wcomm);
+    if(nullptr==wmessage){
```
Comment (boeschf) [inline bindings/communicator_bind.cpp:234]:
```suggestion
    if(!wmessage){
```

File: bindings/message_bind.cpp [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+
+namespace ghex = gridtools::ghex;
+
+using namespace gridtools::ghex::fhex;
+
+namespace gridtools {
+    namespace ghex {
+        namespace fhex {
+            // TODO how to construct the (per-thread) allocators
+            std_allocator_type std_allocator;
```
[the remark this answers is not in the document]
Comment (angainor) [reply bindings/message_bind.cpp:16]: This comes back to the discussion about per-thread pool allocators. Probably I just don't know how to do this correctly.

File: bindings/message_bind.hpp [hunk 29]
```diff
@@ -0,0 +1,15 @@
...
+#define INCLUDED_GHEX_FORTRAN_MESSAGE_BIND_HPP
+
+#include "obj_wrapper.hpp"
+
+#define ALLOCATOR_STD             1
+#define ALLOCATOR_PERSISTENT_STD  2
+#define ALLOCATOR_HOST            3
+#define ALLOCATOR_PERSISTENT_HOST 4
+#define ALLOCATOR_GPU             5
+#define ALLOCATOR_PERSISTENT_GPU  6
```
[the remark this answers is not in the document]
Comment (angainor) [reply bindings/message_bind.hpp:11]: This is again intended for the use with per-thread pool allocator. Happy to discuss.

File: bindings/CMakeLists.txt [hunk 30]
```diff
@@ -1,3 +1,31 @@
+if(EARLY_BUILD)
+  add_executable(sizes sizes.cpp)
+  target_link_libraries(sizes PRIVATE ghexlib)
+  return()
+endif()
+
+message(STATUS "CONFIGURE sizes tool")
```
Comment (mbianco) [inline bindings/CMakeLists.txt]: Do you need this message now?

File: bindings/ghex_defs.hpp [hunk 31]
```diff
@@ -0,0 +1,32 @@
...
+#define DeviceCPU     1
+#define DeviceGPU     2
+
+#define LayoutFieldLast  1
+#define LayoutFieldFirst 2
+
+/* barrier types */
+#define BarrierGlobal 1
+#define BarrierThread 2
+#define BarrierRank   3
```
[the remark this answers is not in the document]
Comment (angainor) [reply bindings/ghex_defs.hpp:24]: done

File: bindings/CMakeLists.txt (hunk 30 above)
Comment (angainor) [reply bindings/CMakeLists.txt]: no. Can remove it if you want.

Comment (angainor) [reply bindings/CMakeLists.txt]: actually, it's not there in the new version.

File: bindings/request_bind.hpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (angainor) [reply bindings/request_bind.hpp]: @boeschf this is now done in an automatic way during compile time, through a simple C++ program that generates an include file with request sizes.

File: bindings/message_bind.hpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (angainor) [reply bindings/message_bind.hpp]: There is only `Host` and `Device` tag now, all others have been removed
